apps/prairielearn/elements/pl-excalidraw/pl-excalidraw.py

```diff
@@ -0,0 +1,218 @@
+
+
+def prepare(element_html: str, data: pl.QuestionData) -> None:
+    element = lxml.html.fragment_fromstring(element_html)
+    required_attrs = []
+    optional_attrs = [
+        ATTR_GRADABLE,
+        ATTR_ANSWER_NAME,
+        ATTR_WIDTH,
+        ATTR_HEIGHT,
+        ATTR_SOURCE_FILE_NAME,
+        ATTR_SOURCE_DIRECTORY,
+    ]
+    pl.check_attribs(element, required_attrs, optional_attrs)
+
+    gradable = pl.get_boolean_attrib(element, ATTR_GRADABLE, True)
+
+    name = pl.get_string_attrib(element, ATTR_ANSWER_NAME, None)
+    if name:
+        pl.check_answers_names(data, name)
+
+    if is_answer_name_required(gradable) and name is None:
+        raise RuntimeError(
+            f"Missing required attribute {ATTR_ANSWER_NAME} (Required when `{ATTR_GRADABLE}` is set)"
+        )
+
+    source_dir = pl.get_string_attrib(element, ATTR_SOURCE_DIRECTORY, ".")
+    if source_dir not in SOURCE_DIRECTORY_MAP:
+        raise RuntimeError(
+            f"{source_dir=} must be one of {list(SOURCE_DIRECTORY_MAP.keys())}"
+        )
+
+
+def load_file_content(element: HtmlElement, data: pl.QuestionData) -> str:
+    file_dir = SOURCE_DIRECTORY_MAP[
+        pl.get_string_attrib(element, ATTR_SOURCE_DIRECTORY, ".")
+    ]
+    file = Path(data["options"][file_dir]) / pl.get_string_attrib(
+        element, ATTR_SOURCE_FILE_NAME
+    )
+    if not file.exists():
+        raise RuntimeError(f"Unknown file path: {file}")
+    return file.read_text(encoding="utf-8")
+
+
+def render(element_html: str, data: pl.QuestionData) -> str:
+    empty_diagram = ""  # pick a default representation
+    element = lxml.html.fragment_fromstring(element_html)
+    drawing_name = pl.get_string_attrib(element, ATTR_ANSWER_NAME, None)
+
+    gradable = pl.get_boolean_attrib(element, ATTR_GRADABLE, True)
+    fresh = drawing_name not in data["submitted_answers"]
+    panel = data["panel"]
+    source_available = pl.has_attrib(element, ATTR_SOURCE_FILE_NAME)
+
+    # the state space (for debugging)
+    matrix = f"{gradable=} {fresh=} {panel=} {source_available=}"
+
+    if is_source_file_name_required(panel, gradable, fresh) and not source_available:
+        raise RuntimeError(f"Missing required attribute `{ATTR_SOURCE_FILE_NAME}`")
+
+    initial_content: str = empty_diagram
+
+    # We sometimes want to render errors without rendering the widget
+    show_widget = True
+
+    # Notes:
+    # 1. We could have simplified the if-else chains below by collapsing branches, but we use
+    # the expanded form to make it easy to compare with the reference logic at
+    # https://github.com/PrairieLearn/PrairieLearn/pull/9900#discussion_r1685163461
+    # 2. Because we already handled raising an error on missing source file, the error cases
+    # and invalid states are both tagged below as `unreachable`
+    unreachable = RuntimeError(f"Unreachable code path reached with state {matrix=}")
+
+    match panel:
+        case "question":
+            if gradable:
```

Comment:
Can you stack these as a match case on the tuple? It results in less indentation and is easier to read:
```python
case (gradeable, fresh, source available):
    match (True, False, True): ...

Reply:
I intentionally chose the non-minified version for human readers to convince themselves of exhaustivity. I also documented this in the _Notes_ section above.

Reply:
This style makes the code substantially longer and harder to read, if for no other reason than the indentation. This also results in using the same exception repeatedly, instead of just once as the default case at the end. Looking at this again, I would actually suggest including the panel in the match case, since this whole section is just devoted to setting a few variables.

Reply:
I do agree that this is difficult to read, and TBH I think there's not a ton of value in structuring code to match a GitHub comment that will inevitably fail to match what this code actually does as it changes in the future.

That said, I think matching on a tuple of three things isn't necessarily easier to read.

A potential middle group: collapse some of the unreachable cases, e.g. replace this:

```py
                if fresh:
                    if source_available:
                        initial_content = load_file_content(element, data)
                    else:
                        raise unreachable
                else:  # submission
                    raise unreachable
```

with this:

```py
                if fresh and source_available:
                    initial_content = load_file_content(element, data)
                else:
                    raise unreachable
```

This reduces indentation/length while still letting one "read" the logic instead of trying to decipher `case (True, False, True)`.

Reply:
Fixed